File: api/src/test/java/org/pmiops/workbench/cohorts/CohortFactoryTest.java
package org.pmiops.workbench.cohorts;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;

import java.sql.Timestamp;
import java.time.Clock;
import java.util.Collections;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.pmiops.workbench.db.model.DbCohort;
import org.pmiops.workbench.db.model.DbCohortReview;
import org.pmiops.workbench.db.model.DbUser;
import org.pmiops.workbench.db.model.DbWorkspace;
import org.pmiops.workbench.model.ReviewStatus;

public class CohortFactoryTest {

  private CohortFactory cohortFactory;

  @BeforeEach
  public void setUp() {
    cohortFactory = new CohortFactoryImpl(Clock.systemUTC());
  }

  @Test
  public void createCohort() {
    org.pmiops.workbench.model.Cohort apiCohort = new org.pmiops.workbench.model.Cohort();
    apiCohort.setDescription("desc");
    apiCohort.setName("name");
    apiCohort.setType("type");
    apiCohort.setCriteria("criteria");

    DbUser user = mock(DbUser.class);

    long workspaceId = 1l;

    DbCohort dbCohort = cohortFactory.createCohort(apiCohort, user, workspaceId);

    assertThat(dbCohort.getDescription()).isEqualTo(apiCohort.getDescription());
    assertThat(dbCohort.getName()).isEqualTo(apiCohort.getName());
    assertThat(dbCohort.getType()).isEqualTo(apiCohort.getType());
    assertThat(dbCohort.getCriteria()).isEqualTo(apiCohort.getCriteria());
    assertThat(dbCohort.getCreator()).isSameAs(user);
    assertThat(dbCohort.getWorkspaceId()).isEqualTo(workspaceId);
  }

  @Test
  public void duplicateCohort() {
    DbCohort originalCohort = new DbCohort();
    originalCohort.setDescription("desc");
    originalCohort.setName("name");
    originalCohort.setType("type");
    originalCohort.setCriteria("criteria");
    originalCohort.setWorkspaceId(1l);
    originalCohort.setCohortReviews(Collections.singleton(mock(DbCohortReview.class)));

    DbUser user = mock(DbUser.class);
    DbWorkspace workspace = mock(DbWorkspace.class);
    doReturn(1l).when(workspace).getWorkspaceId();
    DbCohort dbCohort = cohortFactory.duplicateCohort("new name", user, workspace, originalCohort);

    assertThat(dbCohort.getDescription()).isEqualTo(originalCohort.getDescription());
    assertThat(dbCohort.getName()).isEqualTo("new name");
    assertThat(dbCohort.getType()).isEqualTo(originalCohort.getType());
    assertThat(dbCohort.getCriteria()).isEqualTo(originalCohort.getCriteria());
    assertThat(dbCohort.getCreator()).isSameAs(user);
    assertThat(dbCohort.getWorkspaceId()).isEqualTo(originalCohort.getWorkspaceId());
    assertThat(dbCohort.getCohortReviews()).isNull();
  }

  @Test
  public void duplicateCohortReview() {
    Timestamp now = new Timestamp(Clock.systemUTC().millis());

    DbCohortReview originalCohortReview = new DbCohortReview();
    originalCohortReview.setCohortId(1l);
    originalCohortReview.setCdrVersionId(2l);
    originalCohortReview.setMatchedParticipantCount(3l);
    originalCohortReview.setReviewSize(4l);
    originalCohortReview.setReviewedCount(5l);
    originalCohortReview.setReviewStatusEnum(ReviewStatus.CREATED);

    DbCohort cohort = mock(DbCohort.class);
    doReturn(1l).when(cohort).getCohortId();
    doReturn(now).when(cohort).getCreationTime();
    doReturn(now).when(cohort).getLastModifiedTime();
    DbCohortReview newReview = cohortFactory.duplicateCohortReview(originalCohortReview, cohort);

    assertThat(newReview.getCohortId()).isEqualTo(originalCohortReview.getCohortId());
    assertThat(newReview.getCreationTime()).isEqualTo(now);
    assertThat(newReview.getLastModifiedTime()).isEqualTo(now);
    assertThat(newReview.getCdrVersionId()).isEqualTo(originalCohortReview.getCdrVersionId());
    assertThat(newReview.getMatchedParticipantCount())
        .isEqualTo(originalCohortReview.getMatchedParticipantCount());
    assertThat(newReview.getReviewSize()).isEqualTo(originalCohortReview.getReviewSize());
    assertThat(newReview.getReviewedCount()).isEqualTo(originalCohortReview.getReviewedCount());
    assertThat(newReview.getReviewStatusEnum())
        .isEqualTo(originalCohortReview.getReviewStatusEnum());
  }
}

File: api/src/test/java/org/pmiops/workbench/cohorts/CohortFactoryTest.kt
package org.pmiops.workbench.cohorts

import org.assertj.core.api.Assertions.assertThat
import org.mockito.Mockito.doReturn
import org.mockito.Mockito.mock

import java.sql.Timestamp
import java.time.Clock
import java.util.Collections
import org.junit.Before
import org.junit.Test
import org.pmiops.workbench.db.model.Cohort
import org.pmiops.workbench.db.model.CohortReview
import org.pmiops.workbench.db.model.User
import org.pmiops.workbench.db.model.Workspace
import org.pmiops.workbench.model.ReviewStatus

class CohortFactoryTest {

    private var cohortFactory: CohortFactory? = null

    @Before
    fun setUp() {
        cohortFactory = CohortFactoryImpl(Clock.systemUTC())
    }

    @Test
    fun createCohort() {
        val apiCohort = org.pmiops.workbench.model.Cohort()
        apiCohort.setDescription("desc")
        apiCohort.setName("name")
        apiCohort.setType("type")
        apiCohort.setCriteria("criteria")

        val user = mock(User::class.java)

        val workspaceId = 1L

        val dbCohort = cohortFactory!!.createCohort(apiCohort, user, workspaceId)

        assertThat(dbCohort.description).isEqualTo(apiCohort.getDescription())
        assertThat(dbCohort.name).isEqualTo(apiCohort.getName())
        assertThat(dbCohort.type).isEqualTo(apiCohort.getType())
        assertThat(dbCohort.criteria).isEqualTo(apiCohort.getCriteria())
        assertThat<User>(dbCohort.creator).isSameAs(user)
        assertThat(dbCohort.workspaceId).isEqualTo(workspaceId)
    }

    @Test
    fun duplicateCohort() {
        val originalCohort = Cohort()
        originalCohort.description = "desc"
        originalCohort.name = "name"
        originalCohort.type = "type"
        originalCohort.criteria = "criteria"
        originalCohort.workspaceId = 1L
        originalCohort.cohortReviews = setOf(mock(CohortReview::class.java))

        val user = mock(User::class.java)
        val workspace = mock(Workspace::class.java)
        doReturn(1L).`when`(workspace).workspaceId
        val dbCohort = cohortFactory!!.duplicateCohort("new name", user, workspace, originalCohort)

        assertThat(dbCohort.description).isEqualTo(originalCohort.description)
        assertThat(dbCohort.name).isEqualTo("new name")
        assertThat(dbCohort.type).isEqualTo(originalCohort.type)
        assertThat(dbCohort.criteria).isEqualTo(originalCohort.criteria)
        assertThat<User>(dbCohort.creator).isSameAs(user)
        assertThat(dbCohort.workspaceId).isEqualTo(originalCohort.workspaceId)
        assertThat(dbCohort.cohortReviews).isNull()
    }

    @Test
    fun duplicateCohortReview() {
        val now = Timestamp(Clock.systemUTC().millis())

        val originalCohortReview = CohortReview()
        originalCohortReview.cohortId = 1L
        originalCohortReview.cdrVersionId = 2L
        originalCohortReview.matchedParticipantCount = 3L
        originalCohortReview.reviewSize = 4L
        originalCohortReview.reviewedCount = 5L
        originalCohortReview.reviewStatusEnum = ReviewStatus.CREATED

        val cohort = mock(Cohort::class.java)
        doReturn(1L).`when`(cohort).cohortId
        doReturn(now).`when`(cohort).creationTime
        doReturn(now).`when`(cohort).lastModifiedTime
        val newReview = cohortFactory!!.duplicateCohortReview(originalCohortReview, cohort)

        assertThat(newReview.cohortId).isEqualTo(originalCohortReview.cohortId)
        assertThat(newReview.creationTime).isEqualTo(now)
        assertThat(newReview.lastModifiedTime).isEqualTo(now)
        assertThat(newReview.cdrVersionId).isEqualTo(originalCohortReview.cdrVersionId)
        assertThat(newReview.matchedParticipantCount)
                .isEqualTo(originalCohortReview.matchedParticipantCount)
        assertThat(newReview.reviewSize).isEqualTo(originalCohortReview.reviewSize)
        assertThat(newReview.reviewedCount).isEqualTo(originalCohortReview.reviewedCount)
        assertThat(newReview.reviewStatusEnum)
                .isEqualTo(originalCohortReview.reviewStatusEnum)
    }
}
